core/txpool/txpool.go

```diff
@@ -145,1705 +36,296 @@ const (
-// wrapper around AddRemotes.
-//
-// Deprecated: use AddRemotes
-func (pool *TxPool) AddRemote(tx *types.Transaction) error {
-	errs := pool.AddRemotes([]*types.Transaction{tx})
-	return errs[0]
-}
-
-// addTxs attempts to queue a batch of transactions if they are valid.
-func (pool *TxPool) addTxs(txs []*types.Transaction, local, sync bool) []error {
-	// Filter out known ones without obtaining the pool lock or recovering signatures
-	var (
-		errs = make([]error, len(txs))
-		news = make([]*types.Transaction, 0, len(txs))
-	)
 	for i, tx := range txs {
-		// If the transaction is known, pre-set the error slot
-		if pool.all.Get(tx.Hash()) != nil {
-			errs[i] = ErrAlreadyKnown
-			knownTxMeter.Mark(1)
-			continue
-		}
-		// Exclude transactions with basic errors, e.g invalid signatures and
-		// insufficient intrinsic gas as soon as possible and cache senders
-		// in transactions before obtaining lock
-		if err := pool.validateTxBasics(tx, local); err != nil {
-			errs[i] = err
-			invalidTxMeter.Mark(1)
-			continue
+		for j, subpool := range p.subpools {
+			if subpool.Filter(tx.Tx) {
+				txsets[j] = append(txsets[j], tx)
+				splits[i] = j
+				break
+			}
 		}
-		// Accumulate all unknown transactions for deeper processing
-		news = append(news, tx)
 	}
-	if len(news) == 0 {
-		return errs
+	// Add the transactions split apart to the individual subpools and piece
+	// back the errors into the original sort order.
+	errsets := make([][]error, len(p.subpools))
+	for i := 0; i < len(p.subpools); i++ {
+		errsets[i] = p.subpools[i].Add(txsets[i], local, sync)
 	}
-
-	// Process all the new transaction and merge any errors into the original slice
-	pool.mu.Lock()
-	newErrs, dirtyAddrs := pool.addTxsLocked(news, local)
-	pool.mu.Unlock()
-
-	var nilSlot = 0
-	for _, err := range newErrs {
-		for errs[nilSlot] != nil {
-			nilSlot++
-		}
-		errs[nilSlot] = err
-		nilSlot++
-	}
-	// Reorg the pool internals if needed and return
-	done := pool.requestPromoteExecutables(dirtyAddrs)
-	if sync {
-		<-done
+	errs := make([]error, len(txs))
+	for i, split := range splits {
+		errs[i] = errsets[split][0]
+		errsets[split] = errsets[split][1:]
 	}
 	return errs
 }
 
-// addTxsLocked attempts to queue a batch of transactions if they are valid.
-// The transaction pool lock must be held.
-func (pool *TxPool) addTxsLocked(txs []*types.Transaction, local bool) ([]error, *accountSet) {
-	dirty := newAccountSet(pool.signer)
-	errs := make([]error, len(txs))
-	for i, tx := range txs {
-		replaced, err := pool.add(tx, local)
-		errs[i] = err
-		if err == nil && !replaced {
-			dirty.addTx(tx)
+// Pending retrieves all currently processable transactions, grouped by origin
+// account and sorted by nonce.
+func (p *TxPool) Pending(enforceTips bool) map[common.Address][]*types.Transaction {
+	txs := make(map[common.Address][]*types.Transaction)
+	for _, subpool := range p.subpools {
+		for addr, set := range subpool.Pending(enforceTips) {
```

Comment:
Let's say we have [legacypool, blobpool]

the next nonce of account A is 5

in legacypool, it has tx(nonce=5), tx(nonce=6)
in blobpool, it has tx(nonce=7)

will tx pool somehow aggregate them together as a single executable batch,
or just tx(nonce=5), tx(nonce=6) are returned

Reply:
Currently the invariant is that accounts are exclusive to pools. It is not yet enforced, but should be when blob pool lands. Specifically, if you have txs from an account in one pool, the others should reject it.

Reply:
Otherwise the whole point of multiple subpools becomes moot if they need to agree on state across each other and what to evict / how to track the same things.